zkevm-circuits/src/evm_circuit/execution/stop.rs

```diff
@@ -21,28 +34,192 @@ impl<F: Field> ExecutionGadget<F> for StopGadget<F> {
     const EXECUTION_STATE: ExecutionState = ExecutionState::STOP;
 
     fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let code_size = cb.bytecode_length(cb.curr.state.code_hash.expr());
+        let is_out_of_range =
+            IsZeroGadget::construct(cb, code_size.expr() - cb.curr.state.program_counter.expr());
         let opcode = cb.query_cell();
-        cb.opcode_lookup(opcode.expr(), 1.expr());
+        cb.condition(1.expr() - is_out_of_range.expr(), |cb| {
+            cb.opcode_lookup(opcode.expr(), 1.expr());
+        });
 
-        // Other constraints are ignored now for STOP to serve as a mocking
-        // terminator
+        // We do the responsible opcode check explicitly here because we're not using
+        // the `SameContextGadget` for `STOP`.
+        cb.require_equal(
+            "Opcode should be STOP",
+            opcode.expr(),
+            OpcodeId::STOP.expr(),
+        );
 
-        Self { opcode }
+        // Call ends with STOP must be successful
+        cb.call_context_lookup(false.expr(), None, CallContextFieldTag::IsSuccess, 1.expr());
+
+        let is_to_end_tx = cb.next.execution_state_selector([ExecutionState::EndTx]);
+        cb.require_equal(
+            "Go to EndTx only when is_root",
+            cb.curr.state.is_root.expr() + is_to_end_tx.clone(),
```

Comment:
I believe this comment applies here as well :P https://github.com/privacy-scaling-explorations/zkevm-specs/pull/100#discussion_r859947765